File: src/main/java/me/lokka30/levelledmobs/customdrops/CustomDropItem.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.customdrops;

import java.util.List;
import java.util.Map;
import java.util.UUID;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.util.Utils;
import org.bukkit.Material;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * This holds all the attributes set for a custom drop item
 *
 * @author stumper66
 * @since 2.5.0
 */
public class CustomDropItem extends CustomDropBase {

    public int customModelDataId;
    public float equippedSpawnChance;
    public boolean noMultiplier;
    public boolean onlyDropIfEquipped;
    public boolean equipOnHelmet;
    public String customName;
    public String mobHeadTexture;
    public List<String> lore;
    public @Nullable List<ItemFlag> itemFlags;
    public @Nullable List<String> itemFlagsStrings;
    private boolean hasDamageRange;
    private int damage;
    private int damageRangeMin;
    private int damageRangeMax;
    public boolean equipOffhand;
    public UUID customPlayerHeadId;
    private Material material;
    private ItemStack itemStack;
    public boolean isExternalItem;
    public String externalPluginName;
    public String externalType;
    public String externalItemId;
    public String nbtData;
    public Double externalAmount;
    public Map<String, Object> externalExtras;
    public EnchantmentChances enchantmentChances;

    @SuppressWarnings("unused")
    public CustomDropItem(@NotNull final LevelledMobs levelledMobs) {
        super(levelledMobs.customDropsHandler.customDropsParser.getDefaults());
        setDefaults(levelledMobs.customDropsHandler.customDropsParser.getDefaults());
    }

    public CustomDropItem(@NotNull final CustomDropsDefaults defaults) {
        super(defaults);
        setDefaults(defaults);
    }

    private void setDefaults(@NotNull final CustomDropsDefaults defaults){
        this.customModelDataId = defaults.customModelData;
        this.chance = defaults.chance;
        this.maxLevel = defaults.maxLevel;
        this.minLevel = defaults.minLevel;
        this.groupId = defaults.groupId;
        this.maxDropGroup = defaults.maxDropGroup;
        this.equippedSpawnChance = defaults.equippedSpawnChance;
        this.noMultiplier = defaults.noMultiplier;
        this.onlyDropIfEquipped = defaults.onlyDropIfEquipped;
        this.equipOffhand = defaults.equipOffhand;
        this.equipOnHelmet = defaults.equipOnHelmet;
    }

    public CustomDropItem cloneItem() {
        CustomDropItem copy = null;
        try {
            copy = (CustomDropItem) super.clone();
            copy.itemStack = this.itemStack.clone();
        } catch (final Exception e) {
            e.printStackTrace();
        }

        return copy;
    }

    public boolean setDamageRangeFromString(final String numberOrNumberRange) {
        if (numberOrNumberRange == null || numberOrNumberRange.isEmpty()) {
            return false;
        }

        if (!numberOrNumberRange.contains("-")) {
            if (!Utils.isInteger(numberOrNumberRange)) {
                return false;
            }

            this.damage = Integer.parseInt(numberOrNumberRange);
            this.hasDamageRange = false;
            this.damageRangeMax = 0;
            this.damageRangeMin = 0;
            return true;
        }

        final String[] nums = numberOrNumberRange.split("-");
        if (nums.length != 2) {
            return false;
        }

        if (!Utils.isInteger(nums[0].trim()) || !Utils.isInteger(nums[1].trim())) {
            return false;
        }
        this.damageRangeMin = Integer.parseInt(nums[0].trim());
        this.damageRangeMax = Integer.parseInt(nums[1].trim());
        this.hasDamageRange = true;

        return true;
    }

    public int getDamage() {
        return this.damage;
    }

    public void setDamage(final int damage) {
        this.damage = damage;
        this.hasDamageRange = false;
    }

    public int getDamageRangeMin() {
        return this.damageRangeMin;
    }

    public int getDamageRangeMax() {
        return this.damageRangeMax;
    }

    public boolean getHasDamageRange() {
        return this.hasDamageRange;
    }

    public void setMaterial(final Material material) {
        this.material = material;
        this.itemStack = new ItemStack(this.material, 1);
    }

    public @NotNull String getDamageAsString() {
        if (this.hasDamageRange) {
            return String.format("%s-%s", this.damageRangeMin, this.damageRangeMax);
        } else {
            return String.valueOf(this.damage);
        }
    }

    public Material getMaterial() {
        return this.material;
    }

    public ItemStack getItemStack() {
        return itemStack;
    }

    public void setItemStack(final @NotNull ItemStack itemStack) {
        this.itemStack = itemStack;
        this.material = itemStack.getType();
    }

    public String toString() {
        return String.format("%s, amount: %s, chance: %s, equipped: %s",
            this.material.name(), this.getAmountAsString(), this.chance, this.equippedSpawnChance);
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/customdrops/CustomDropItem.kt
package io.github.arcaneplugins.levelledmobs.customdrops

import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.util.Utils
import org.bukkit.Material
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack

/**
 * This holds all the attributes set for a custom drop item
 *
 * @author stumper66
 * @since 2.5.0
 */
class CustomDropItem() : CustomDropBase(
        LevelledMobs.instance.customDropsHandler.customDropsParser.defaults),
        Cloneable {
    var customModelDataId = 0
    var amountFormula: String? = null
    var equippedChance: SlidingChance? = null
    var noMultiplier = false
    var onlyDropIfEquipped = false
    var equipOnHelmet = false
    var customName: String? = null
    var lore: MutableList<String>? = null
    var itemFlags: MutableList<ItemFlag>? = null
    var itemFlagsStrings: MutableList<String>? = null
    var allowedList: MutableList<String>? = null
    var excludedList: MutableList<String>? = null
    var hasDamageRange = false
        private set
    var damageRangeMin = 0
    var damageRangeMax = 0
    var minItems = 0
    var maxItems = 0
    var equipOffhand = false
    private var _material: Material? = null
    var isExternalItem = false
    var externalPluginName: String? = null
    var externalType: String? = null
    var externalItemId: String? = null
    var nbtData: String? = null
    var externalAmount: Double? = null
    var externalExtras: MutableMap<String, Any>? = null
    var enchantmentChances: EnchantmentChances? = null

    constructor(defaults: CustomDropsDefaults) : this() {
        setDefaults(defaults)
    }

    private fun setDefaults(defaults: CustomDropsDefaults) {
        this.customModelDataId = defaults.customModelData
        if (this.chance != null)
            this.chance!!.setFromInstance(defaults.chance)
        else
            this.chance = defaults.chance
        this.maxLevel = defaults.maxLevel
        this.minLevel = defaults.minLevel
        this.groupId = defaults.groupId
        this.maxDropGroup = defaults.maxDropGroup
        if (this.equippedChance != null)
            this.equippedChance!!.setFromInstance(defaults.equippedChance)
        else
            this.equippedChance = defaults.equippedChance
        this.noMultiplier = defaults.noMultiplier
        this.onlyDropIfEquipped = defaults.onlyDropIfEquipped
        this.equipOffhand = defaults.equipOffhand
        this.equipOnHelmet = defaults.equipOnHelmet
    }

    fun cloneItem(): CustomDropItem? {
        var copy: CustomDropItem? = null
        try {
            copy = super.clone() as CustomDropItem
            copy.itemStack = itemStack!!.clone()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return copy
    }

    fun setDamageRangeFromString(numberOrNumberRange: String?): Boolean {
        if (numberOrNumberRange.isNullOrEmpty()) {
            return false
        }

        if (!numberOrNumberRange.contains("-")) {
            if (!Utils.isInteger(numberOrNumberRange)) {
                return false
            }

            this.damage = numberOrNumberRange.toInt()
            this.hasDamageRange = false
            this.damageRangeMax = 0
            this.damageRangeMin = 0
            return true
        }

        val nums = numberOrNumberRange.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        if (nums.size != 2) {
            return false
        }

        if (!Utils.isInteger(nums[0].trim { it <= ' ' }) || !Utils.isInteger(
                nums[1].trim { it <= ' ' })
        ) {
            return false
        }
        this.damageRangeMin = nums[0].trim { it <= ' ' }.toInt()
        this.damageRangeMax = nums[1].trim { it <= ' ' }.toInt()
        this.hasDamageRange = true

        return true
    }

    var damage = 0
        set(value) {
            field = value
            this.hasDamageRange = false
        }

    var material: Material
        get() = this._material!!
        set(value) {
            this._material = value
            this.itemStack = ItemStack(_material!!, 1)
        }

    fun getDamageAsString(): String {
        return if (this.hasDamageRange) {
            "$damageRangeMin-$damageRangeMax"
        } else {
            damage.toString()
        }
    }

    var itemStacks: MutableList<ItemStack>? = null
        set(value) {
            field = value
            if (itemStacks != null && itemStacks!!.isNotEmpty()) {
                this.itemStack = itemStacks!![0]
                this.material = itemStack!!.type
            }
        }
        get() {
            if (field != null && field!!.isNotEmpty())
                return field

            if (itemStack == null) return null
            return mutableListOf(itemStack!!)
        }

    var itemStack: ItemStack? = null
        set(value) {
            field = value
            _material = value!!.type
        }

    override fun toString(): String {
        return "${material.name}, amount: $amountAsString, chance: $chance, equipped: $equippedChance"
    }
}